src/clojure/nrepl/socket.clj

```diff
@@ -0,0 +1,222 @@
+(ns nrepl.socket
+  "Compatibility layer for java.io vs java.nio sockets to allow an
+  incremental transition to nio, since the JDK's filesystem sockets
+  don't support the java.io socket interface, and we can't use the
+  compatibility layer for bidirectional read and write:
+  https://bugs.openjdk.java.net/browse/JDK-4509080."
+  (:require
+   [clojure.java.io :as io]
+   [nrepl.misc :refer [log]])
+  (:import
+   (clojure.lang Reflector)
+   (java.io BufferedInputStream BufferedOutputStream File)
+   (java.net InetSocketAddress ServerSocket Socket StandardProtocolFamily URI)
+   (java.nio ByteBuffer)
+   (java.nio.file Path)
+   (java.nio.channels Channels ClosedChannelException
+                      ServerSocketChannel SocketChannel)))
+
+(def orig-warn-on-reflection *warn-on-reflection*)
+
+(defmacro find-class [full-path]
+  `(try
+     (import ~full-path)
+     (catch ClassNotFoundException ex#
+       nil)))
+
+;;; InetSockets (TCP)
+
+(defn inet-socket [bind port]
+  (let [port (or port 0)
+        addr (fn [^String bind ^Integer port] (InetSocketAddress. bind port))
+        ;; We fallback to 127.0.0.1 instead of to localhost to avoid
+        ;; a dependency on the order of ipv4 and ipv6 records for
+        ;; localhost in /etc/hosts
+        bind (or bind "127.0.0.1")]
+    (doto (ServerSocket.)
+      (.setReuseAddress true)
+      (.bind (addr bind port)))))
+
+;; Unix domain sockets
+
+(def ^Class junixsocket-address-class
+  (find-class 'org.newsclub.net.unix.AFUNIXSocketAddress))
+
+(def ^Class junixsocket-server-class
+  (find-class 'org.newsclub.net.unix.AFUNIXServerSocket))
+
+(def ^Class jdk-unix-address-class
+  (find-class 'java.net.UnixDomainSocketAddress))
+
+(def ^Class jdk-unix-server-class
+  (find-class 'java.nio.channels.ServerSocketChannel))
+
+(def ^:private test-junixsocket?
+  ;; Make it possible to test junixsocket even when JDK >= 16
+  (= "true" (System/getProperty "nrepl.test.junixsocket")))
+
+(def ^:const unix-domain-flavor
+  (cond
+    test-junixsocket? (do
+                        (assert junixsocket-address-class)
+                        (assert junixsocket-server-class)
+                        (binding [*out* *err*]
+                          (println "nrepl.test: insisting on junixsocket support"))
+                        :junixsocket)
+    (and jdk-unix-address-class jdk-unix-server-class) :jdk
+    (and junixsocket-address-class junixsocket-server-class) :junixsocket
+    :else nil))
+
+(defn unix-socket-address
+  "Returns a filesystem socket address for the given path string."
+  [^String path]
+  (case unix-domain-flavor
+    :jdk
+    (Reflector/invokeStaticMethod jdk-unix-address-class "of"
```

Comment:
Not sure if lang.Reflector is a public API. Did you consider using java's vanilla reflection API?

Reply:
I have now :)
(Fix pending.)